benchmarks/benchmark_base.py

```diff
@@ -0,0 +1,453 @@
+        )
+        config.plasma.nlte_solver = "root"
+        return config
+
+    @property
+    def tardis_model_config_nlte_lu(self):
+        config = Configuration.from_yaml(
+            f"{self.example_configuration_dir}/tardis_configv1_nlte.yml"
+        )
+        config.plasma.nlte_solver = "lu"
+        return config
+
+    @property
+    def example_configuration_dir(self):
+        return self.get_absolute_path("tardis/io/configuration/tests/data")
+
+    @property
+    def hdf_file_path(self):
+        # TODO: Delete this files after ASV runs the benchmarks.
+        #       ASV create a temporal directory in runtime per test: `tmpiuxngvlv`.
+        #       The ASV and ASV_Runner, not has some way to get this temporal directory.
+        #       The idea is use this temporal folders to storage this created temporal file.
+        _, path = mkstemp('-tardis-benchmark-hdf_buffer-test.hdf')
+        return path
+
+    def create_temporal_file(self, suffix=None):
+        # TODO: Delete this files after ASV runs the benchmarks.
+        #       ASV create a temporal directory in runtime per test: `tmpiuxngvlv`.
+        #       The ASV and ASV_Runner, not has some way to get this temporal directory.
+        #       The idea is use this temporal folders to storage this created temporal file.
+        suffix_str = '' if suffix is None else f'-{suffix}'
+        _, path = mkstemp(suffix_str)
+        return path
+
+    @property
+    def gamma_ray_simulation_state(self):
+        self.gamma_ray_config.model.structure.velocity.start = 1.0 * u.km / u.s
+        self.gamma_ray_config.model.structure.density.rho_0 = 5.0e2 * u.g / u.cm ** 3
+        self.gamma_ray_config.supernova.time_explosion = 150 * u.d
+
+        return SimulationState.from_config(
+            self.gamma_ray_config, atom_data=self.atomic_dataset
+        )
+
+    @property
+    def gamma_ray_config(self):
+        yml_path = (
+            f"{self.example_configuration_dir}/tardis_configv1_density_exponential_nebular_multi_isotope.yml"
+        )
+
+        return config_reader.Configuration.from_yaml(yml_path)
+
+    @property
+    def example_model_file_dir(self):
+        return self.get_absolute_path("tardis/io/model/readers/tests/data")
+
+    @property
+    def kurucz_atomic_data(self) -> AtomData:
+        return deepcopy(self.atomic_dataset)
+
+    @property
+    def example_csvy_file_dir(self):
+        return self.get_absolute_path("tardis/model/tests/data/")
+
+    @property
+    def simulation_verysimple(self):
+        atomic_data = deepcopy(self.atomic_dataset)
+        sim = Simulation.from_config(self.config_verysimple, atom_data=atomic_data)
+        sim.iterate(4000)
+        return sim
+
+    @property
+    def config_verysimple(self):
+        return Configuration.from_yaml(
+            f"{self.example_configuration_dir}/tardis_configv1_verysimple.yml"
+        )
+
+    class RegressionData:
```

Comment:
That's very useful information. My intent with the unit test first objective was that our unit tests provide suggested input for functions that make benchmarking them easier for applicants without prior TARDIS knowledge. If this leads to extensive code duplication, then more "integrated" benchmarks may be preferable, that measure multiple functions at once.

Reply:
I don't know the context here but since I was mentioned I will say that in general, benchmarks and tests are very similar in the way they work, so they can probably share some common frameworks. But I would caution against saying that they are the same thing in general. Tests and benchmarks have different goals and trying to squeeze both into one is a bad idea. A good test may be a bad benchmark and a good benchmark may be a bad test. 

If anything, a benchmark is like a "test" against performance instead of functionality, but even then, that leads to different requirements. Just as an example, asv benchmarks should work against every historical version of the code, whereas a test only needs to run against the latest version.

Reply:
Thanks for your insight, Aaron!

Reply:
Thanks, Aron!

Reply:
I feel lost in these approaches because I don't know what the best setup that provides a useful result for the benchmark.

I was reviewing the article about [Pandas regression using ASV benchmarking](http://tomaugspurger.net/posts/performance-regressions/) that @epassaro shared. Then with all of this information I have these comments:

I was thinking carefully and came to the conclusion that we can learn from those who have already walked the path. So I took on the task of searching the Pandas repository to see how they handle benchmarks. This is the [Pandas repository regarding benchmarks](https://github.com/pandas-dev/pandas/tree/main/asv_bench/benchmarks) versus [unit testing for memory_usage](https://github.com/search?q=repo%3Apandas-dev%2Fpandas+path%3Atest_+test_memory_usage&type=code), but I didn't find a pattern to follow. Apparently, the [benchmarks are very specific and simple](https://github.com/pandas-dev/pandas/blob/2fbabb11db42d3a109cd13e3377327c8c1834213/asv_bench/benchmarks/frame_methods.py#L794). Normally, they test a method with parameters; however, I didn't see many benchmarks with parameterization. Perhaps a good idea is to schedule a meeting with the Pandas maintainers to see if they will give us 30 minutes or an hour to give us a lecture on how to implement our benchmarks.

The second option is to apply all the approaches and learning along the way for months or years. The approaches could be all of these three together:
- Create benchmarks based on the unit tests.
- Create some specific unit tests for some specific methods.
- As you mentioned, Andrew, create benchmarks with flows of functionality (“integrated” benchmarks).